pkg/logcli/query/tail.go

```diff
@@ -44,9 +42,43 @@ func (q *Query) TailQuery(delayFor time.Duration, c client.Client, out output.Lo
 		log.Println("Print only labels key:", color.RedString(strings.Join(q.ShowLabelsKey, ",")))
 	}
 
+	tailResponse := new(loghttp.TailResponse)
+	lastReceivedTimestamp := q.Start
+
 	for {
 		err := unmarshal.ReadTailResponseJSON(tailResponse, conn)
 		if err != nil {
+			// Check if the websocket connection closed unexpectedly. If so, retry.
+			// The connection might close unexpectedly if the querier handling the tail request
+			// in Loki stops running. The following error would be printed:
+			// "websocket: close 1006 (abnormal closure): unexpected EOF"
+			if websocket.IsCloseError(err, websocket.CloseAbnormalClosure) {
+				log.Printf("Remote websocket connection closed unexpectedly (%+v). Connecting again.", err)
+
+				// Close previous connection. If it fails to close the connection it should be fine as it is already broken.
+				if err = conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")); err != nil {
+					log.Printf("Error closing websocket: %+v", err)
+				}
+
+				// Try to re-establish the connection up to 5 times.
+				for retry := 4; retry >= 0; retry-- {
```

Comment:
We could use `github.com/grafana/dskit/backoff` here for the retry I think.

For example how to use it, recently Travis added it on promtail gcp target.
https://github.com/grafana/loki/pull/8281

Reply:
That's super cool! Changed to use it with [fe8017e](https://github.com/grafana/loki/pull/8374/commits/fe8017ee8f9557e5f3fada48fd366ceb3fceec5f). Tested and works like a charm. Thanks for the tip 💯